zpool: Add zpool status -vv error ranges

tonyhutter opened the pull request:
### Motivation and Context
Print error byte ranges with `zpool status -vv`

### Description
Print the byte error ranges with 'zpool status -vv'.  This works with all the normal zpool status formatting flags: `-p`, `-j`, `--json-int`

In addition:

- Move range_tree/btree to common userspace/kernel code.
- Modify ZFS_IOC_OBJ_TO_STATS ioctl to optionally return "extended" object stats.
- Let zinject corrupt zvol data.
- Add test case.

This commit takes code from these PRs: #17502 #9781 #8902

### How Has This Been Tested?
Test case added

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

asomers committed:
zpool: Add zpool status -vv error ranges

Print the byte error ranges with 'zpool status -vv'.  This works
with the normal zpool status formatting flags (-p, -j, --json-int).

In addition:

- Modify ZFS_IOC_OBJ_TO_STATS ioctl to optionally return "extended"
  object stats.
- Let zinject corrupt zvol data.
- Add test case.
- Add DIOCGPHYSPATH geom attribute to FreeBSD zvols.  This works
  basically the same as Linux's BLKZNAME ioctl, and returns the
  dataset name for a zvol (like "tank/vol").  This is used by zinject.

This commit takes code from these PRs: #17502 #9781 #8902

Signed-off-by: Tony Hutter <hutter2@llnl.gov>
Co-authored-by:: Alan Somers <asomers@gmail.com>
Co-authored-by: TulsiJain <tulsi.jain@delphix.com>

tonyhutter commented:
Sample output:
```
$ zpool status -vv

  pool: testpool                                                     
 state: ONLINE                                                       
status: One or more devices has experienced an error resulting in data
    corruption.  Applications may be affected.                       
action: Restore the file in question if possible.  Otherwise restore the
    entire pool from backup.                                         
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-8A       
  scan: scrub repaired 0B in 00:00:00 with 3 errors on Tue Oct 21 17:22:20 2025
config:                                                              
                                                                     
    NAME        STATE     READ WRITE CKSUM                           
    testpool    ONLINE       0     0     0                           
      loop0     ONLINE       3     0    18                           
                                                                     
errors: Permanent errors have been detected in the following files:  
                                                                     
        <metadata>:<0x1> (no ranges)                                 
        /testpool/4k/4k_file1 0-4.00K                                
        /testpool/4k/4k_file2 0-4.00K,16K-20.0K,24K-28.0K            
        /testpool/1m/1m_file 1M-2.00M                                
        testpool/testvol:<0x1> 3.91M-3.91M,4.30M-4.30M    
```
```
$ zpool status -vvjp --json-int | jq
...
      "errors": {
        "<metadata>:<0x1>": {
          "name": "<metadata>:<0x1>",
          "object": 1,
          "dataset": 0
        },
        "/testpool/4k/4k_file1": {
          "object_type": "ZFS plain file",
          "ranges": [
            {
              "start_byte": 0,
              "end_byte": 4095
            }
          ],
          "name": "/testpool/4k/4k_file1",
          "object": 2,
          "dataset": 262,
          "block_size": 4096
        },
        "/testpool/4k/4k_file2": {
          "object_type": "ZFS plain file",
          "ranges": [
            {
              "start_byte": 0,
              "end_byte": 4095
            },
            {
              "start_byte": 16384,
              "end_byte": 20479
            },
            {
              "start_byte": 24576,
              "end_byte": 28671
            }
          ],
          "name": "/testpool/4k/4k_file2",
          "object": 128,
          "dataset": 262,
          "block_size": 4096
        },
        "/testpool/1m/1m_file": {
          "object_type": "ZFS plain file",
          "ranges": [
            {
              "start_byte": 1048576,
              "end_byte": 2097151
            }
          ],
          "name": "/testpool/1m/1m_file",
          "object": 2,
          "dataset": 270,
          "block_size": 1048576
        },
        "testpool/testvol:<0x1>": {
          "object_type": "zvol",
          "ranges": [
            {
              "start_byte": 4096000,
              "end_byte": 4100095
            },
            {
              "start_byte": 4505600,
              "end_byte": 4509695
            }
          ],
          "name": "testpool/testvol:<0x1>",
          "object": 1,
          "dataset": 278,
          "block_size": 4096
        }
      }
    }
  }
}
```

tonyhutter commented:
Just the filenames (JSON):
```
$ zpool status -vj
...
      "errors": {
        "<metadata>:<0x1>": {
          "name": "<metadata>:<0x1>"
        },
        "/testpool/4k/4k_file1": {
          "name": "/testpool/4k/4k_file1"
        },
        "/testpool/4k/4k_file2": {
          "name": "/testpool/4k/4k_file2"
        },
        "/testpool/1m/1m_file": {
          "name": "/testpool/1m/1m_file"
        },
        "testpool/testvol:<0x1>": {
          "name": "testpool/testvol:<0x1>"
        }
      }
    }
  }
```

behlendorf commented on cmd/zinject/translate.c:
```diff
@@ -75,11 +75,24 @@ compress_slashes(const char *src, char *dest)
+
+	/* Resolve symlinks to /dev/zd* device */
+	if (realpath(inpath, buf) != NULL)
+		if (strncmp(buf, "/dev/zd", 7) == 0)
```
```suggestion
		if (strncmp(buf, "/dev/" ZVOL_DEV_NAME, 7) == 0)
```

behlendorf commented on cmd/zinject/translate.c:
```diff
@@ -98,6 +111,47 @@ parse_pathname(const char *inpath, char *dataset, char *relpath,
+	if (path_is_zvol(inpath)) {
+		int fd;
+		char *slash;
+		int rc;
```
nit: missing newline after the declarations.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -10341,13 +10383,15 @@ print_checkpoint_status(pool_checkpoint_stat_t *pcs)
 	    space_buf);
 }
 
+
```
nit: stray extra newline

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -10367,9 +10411,49 @@ print_error_log(zpool_handle_t *zhp)
-		(void) printf("%7s %s\n", "", pathname);
+		if (last_pathname == NULL ||
+		    0 != strncmp(pathname, last_pathname, len)) {
+			last_pathname = strdup(pathname);
```
This will be leaked when the pathname changes.

behlendorf commented on lib/libzfs/libzfs_pool.c:
```diff
@@ -4613,6 +4621,182 @@ zpool_add_propname(zpool_handle_t *zhp, const char *propname)
 }
 
+/*
+ * Given an properties nvlist like:
```
```suggestion
 * Given a properties nvlist like:
```

behlendorf commented on lib/librange_tree/Makefile.am:
```diff
@@ -0,0 +1,7 @@
+librange_tree_la_CFLAGS  = $(AM_CFLAGS) $(KERNEL_CFLAGS) $(LIBRARY_CFLAGS)
+# librange_tree_la_CFLAGS += -fvisibility=hidden
```
Shouldn't be commented out.

behlendorf commented on lib/librange_tree/Makefile.am:
```diff
@@ -0,0 +1,7 @@
+noinst_LTLIBRARIES += librange_tree.la
+
+nodist_librange_tree_la_SOURCES = module/zfs/range_tree.c
+nodist_EXTRA_librange_tree_la_SOURCES = module/zfs/btree.c
```
Why is this needed?

behlendorf commented on lib/libspl/include/sys/sysmacros.h:
```diff
@@ -30,7 +30,7 @@
 
 #include <stdint.h>
-
+#include <sys/param.h>
```
I believe you'll want `<limits.h>` for `CHAR_BIT` and the highbit64/lowbit64 changes.

behlendorf commented on lib/libzfs/Makefile.am:
```diff
@@ -65,6 +67,7 @@ libzfs_la_LIBADD = \
 
 libzfs_la_LIBADD += -lrt -lm $(LIBCRYPTO_LIBS) $(ZLIB_LIBS) $(LIBFETCH_LIBS) $(LTLIBINTL)
 
+# libzfs_la_LDFLAGS = -pthread -lrange_tree -lbtree
```
Can be dropped.

behlendorf reviewed (commented):
I only managed to make it through the first patch but I wanted to post a first round of feedback.  I like the direction this is going!

tonyhutter commented:
> @behlendorf let me move over the btree/range_tee commit to another PR as you suggested, and I'll address your comments in there.

Here's the btree/range_tree PR https://github.com/openzfs/zfs/pull/18133.  I believe it addresses all your comments. We'll put this PR on ice until #18133 is merged.

behlendorf commented on cmd/zinject/translate.c:
```diff
@@ -75,17 +76,53 @@ compress_slashes(const char *src, char *dest)
+	if (S_ISLNK(sb.st_mode) && realpath(inpath, buf) != NULL) {
+		/* Resolve symlinks to /dev/zd* device */
+		if (strncmp(buf, devname, strlen(devname)) == 0)
+			return (B_TRUE);
```
At least on linux the `BLKZNAME` ioctl is wired up which could be used to more reliably check if this is a zvol.  See `udev/zvol_id.c`.

behlendorf commented on module/zfs/zfs_ioctl.c:
```diff
@@ -286,6 +286,14 @@ static int zfs_fill_zplprops_root(uint64_t, nvlist_t *, nvlist_t *,
+static inline boolean_t zfs_cmd_has_dst_nvlist(zfs_cmd_t *zc)
+{
+	return (zc->zc_nvlist_dst != 0 && zc->zc_nvlist_dst_size != 0);
+}
```
Let's remove the helper function.  I'm not against adding a helper for this, but if we're going to do it we should apply it to all of the other `zfs_ioc_*` functions which support an optional dst nvlist.  And doing that should be a follow up PR.

behlendorf commented on module/zfs/zfs_ioctl.c:
```diff
@@ -1946,10 +1964,23 @@ zfs_ioc_obj_to_stats(zfs_cmd_t *zc)
+	if (zfs_cmd_has_dst_nvlist(zc)) {
+		VERIFY0(nvlist_alloc(&nv, NV_UNIQUE_NAME, 0));
+	} else
+		nv = NULL;
```
I'd drop the else and initialize `nv = NULL` above.

behlendorf commented on module/zfs/zfs_znode.c:
```diff
@@ -280,9 +280,54 @@ zfs_obj_to_path(objset_t *osp, uint64_t obj, char *buf, int len)
+	dmu_object_info_t doi;
+	sa_object_info(hdl, &doi);
+
+	VERIFY0(nvlist_add_uint32(nv, ZFS_OBJ_STAT_DATA_BLOCK_SIZE,
```
```suggestion
	fnvlist_add_uint32(nv, ZFS_OBJ_STAT_DATA_BLOCK_SIZE,
```

is identical and a little more readable.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool_status/zpool_status_-v.ksh:
```diff
@@ -0,0 +1,182 @@
+#
+# NOTE: for whatever reason, reading from the 1st megabyte of the zvol, with
+# any block size, will not produce an error.  If you read past the 1st megabyte
+# it will.
```
The first `zvol_prefetch_bytes` of a zvol are prefetched on open to speed up things like `blkid`.  Setting `zvol_prefetch_bytes=0` should prevent this.

akashb-22 reviewed (approved):
LGTM.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2591,7 +2594,7 @@ typedef struct status_cbdata {
 	int		cb_namewidth;
 	boolean_t	cb_allpools;
-	boolean_t	cb_verbose;
+	int		cb_verbosity;
```
nit: should be moved before `cb_allpools` to allow the `boolean_t`'s to pack tightly in the struct when possible.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -9647,57 +9679,96 @@ spares_nvlist(zpool_handle_t *zhp, status_cbdata_t *cb, nvlist_t *nv,
+		nvlist_t *nv;
+		uint64_t dsobj, obj;
+
+		verify(nvpair_value_nvlist(elem, &nv) == 0);
```
```suggestion
		nv = fnvpair_value_nvlist(elem);
```

alek-p commented on cmd/zpool/zpool_main.c:
```diff
@@ -9647,57 +9679,96 @@ spares_nvlist(zpool_handle_t *zhp, status_cbdata_t *cb, nvlist_t *nv,
-					    &nv) == 0);
-					verify(nvlist_lookup_uint64(nv,
-					    ZPOOL_ERR_DATASET, &dsobj) == 0);
-					verify(nvlist_lookup_uint64(nv,
```
would be nice to keep ZPOOL_CONFIG_ERRCOUNT in the output

tonyhutter replied:
The internal error count is kind of misleading, as there can be duplicate or overlapping entries.  You can do some `jq` queries on the JSON error list to get the number of error ranges.

alek-p commented on lib/libzfs/libzfs_pool.c:
```diff
@@ -4747,6 +4936,8 @@ zpool_get_errlog(zpool_handle_t *zhp, nvlist_t **nverrlistp)
 	zbookmark_phys_t *zb = buf + zc.zc_nvlist_dst_size;
 	uint64_t zblen = buflen - zc.zc_nvlist_dst_size;
+	if (zblen == 0)
+		return (0); /* nothing to do */
```
think we need to `free(buf);` here

alek-p commented on lib/libzfs/libzfs_pool.c:
```diff
@@ -4697,6 +4699,192 @@ zpool_add_propname(zpool_handle_t *zhp, const char *propname)
+	if (is_zvol)
+		*type_str = "zvol";
+	else
+		*type_str = fnvlist_lookup_string(nv, ZFS_OBJ_STAT_TYPE_STR);
```
do we need to copy the string and `nvlist_free(nv)` here?

tonyhutter replied:
good catch, I will add that

alek-p commented on cmd/zpool/zpool_main.c:
```diff
@@ -9647,57 +9679,96 @@ spares_nvlist(zpool_handle_t *zhp, status_cbdata_t *cb, nvlist_t *nv,
+	nvlist_t **ranges;
+	uint_t count;
+
+	if (zpool_get_errlog(zhp, &nverrlist) != 0)
```
should this be gated by `cb_verbosity > 0` like before?

alek-p commented on cmd/zpool/zpool_main.c:
new output might want to keep the errno reporting?

alek-p commented on cmd/zpool/zpool_main.c:
```diff
@@ -10394,8 +10467,48 @@ print_error_log(zpool_handle_t *zhp)
 		zpool_obj_to_path(zhp, dsobj, obj, pathname, len);
-		(void) printf("%7s %s\n", "", pathname);
+		if (last_pathname[0] == '\0' ||
+		    strncmp(pathname, last_pathname, len) != 0) {
```
should this be `strncmp(pathname, last_pathname, MAXPATHLEN)`

alek-p commented on module/zfs/zfs_znode.c:
```diff
@@ -280,9 +280,47 @@ zfs_obj_to_path(objset_t *osp, uint64_t obj, char *buf, int len)
+	    zfs_obj_type_name(doi.doi_type));
+	fnvlist_add_uint64(nv, ZFS_OBJ_STAT_BONUS_TYPE, doi.doi_bonus_type);
+	fnvlist_add_string(nv, ZFS_OBJ_STAT_BONUS_TYPE_STR,
+	    zfs_obj_type_name(doi.doi_bonus_type));
```
should we add `fnvlist_add_uint64(nv, ZFS_OBJ_STAT_BONUS_SIZE, doi.doi_bonus_size);` here?

alek-p commented on module/zfs/zfs_ioctl.c:
```diff
@@ -2166,10 +2176,19 @@ zfs_ioc_obj_to_stats(zfs_cmd_t *zc)
+	    sizeof (zc->zc_value), nv);
 	dmu_objset_rele_flags(os, B_TRUE, FTAG);
 
+	if (nv != NULL) {
```
this should probably be `if (error == 0 && nv != NULL)`

tonyhutter replied:
except `zfs_obj_to_stats()` can return an error (with `nv` allocated).  I think this will work better to preserve an earlier `error` value:
```diff
        if (nv != NULL) {
-               error = put_nvlist(zc, nv);
+               int tmp = put_nvlist(zc, nv);
+               if (error == 0)
+                       error = tmp;
+
                nvlist_free(nv);
        }
```

alek-p replied:
this is marked as resovled but the above wasn't implemented?

behlendorf replied:
Thanks.  I've added a commit which makes this change.

alek-p commented on lib/libzfs/libzfs_pool.c:
```diff
@@ -5209,6 +5411,69 @@ zpool_events_seek(libzfs_handle_t *hdl, uint64_t eid, int zevent_fd)
+	}
+
+	zcmd_read_dst_nvlist(zhp->zpool_hdl, &zc, &nv);
+
```
need a `zcmd_free_nvlists()` here?

tonyhutter commented:
@alek-p my latest couple of pushes should address your comments.  I also added ZPOOL_CONFIG_ERRCOUNT back to the JSON output and renamed the new error tree from 'errors' -> 'errlist' to keep the old naming conventions.

alek-p commented on lib/libzfs/libzfs_pool.c:
```diff
@@ -5209,6 +5411,70 @@ zpool_events_seek(libzfs_handle_t *hdl, uint64_t eid, int zevent_fd)
+		if (errno == ENOMEM) {
+			zcmd_expand_dst_nvlist(zhp->zpool_hdl, &zc);
+		} else {
+			return (NULL);
```
didn't see it last time but looks like we also need `zcmd_free_nvlists(&zc);` here as well

behlendorf replied:
Fixed.

behlendorf committed:
Address review feedback

This commit should be squashed before merging.

Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>

tonyhutter commented:
My push from 4 days ago addresses all the Copilot changes.
